src/policy/marksweepspace/native_ms/global.rs

```diff
@@ -0,0 +1,369 @@
+use std::sync::Arc;
+
+use atomic::Ordering;
+
+use crate::{
+    policy::{marksweepspace::native_ms::*, sft::GCWorkerMutRef, space::SpaceOptions},
+    scheduler::{GCWorkScheduler, GCWorker},
+    util::{
+        copy::CopySemantics,
+        heap::{
+            layout::heap_layout::{Mmapper, VMMap},
+            FreeListPageResource, HeapMeta, VMRequest,
+        },
+        metadata::{
+            self,
+            side_metadata::{SideMetadataContext, SideMetadataSpec},
+            MetadataSpec,
+        },
+        ObjectReference,
+    },
+    vm::VMBinding,
+};
+
+use crate::plan::ObjectQueue;
+use crate::plan::VectorObjectQueue;
+use crate::policy::sft::SFT;
+use crate::policy::space::{CommonSpace, Space};
+use crate::util::constants::LOG_BYTES_IN_PAGE;
+use crate::util::heap::chunk_map::*;
+use crate::util::linear_scan::Region;
+use crate::util::VMThread;
+use crate::vm::ObjectModel;
+use std::sync::Mutex;
+
+/// The result for `MarkSweepSpace.acquire_block()`. `MarkSweepSpace` will attempt
+/// to allocate from abandoned blocks first. If none found, it will get a new block
+/// from the page resource.
+pub enum BlockAcquireResult {
+    /// A new block we just acquired from the page resource
+    Fresh(Block),
```

Comment:
Consider adding a special case for space exhaustion, because `acquire_block` may fail.
```suggestion
    /// No block is available.
    Exhausted,
    /// A new block we just acquired from the page resource
    Fresh(Block),
```

Reply:
Done